### PR TITLE
CB-13472 - Validate for stopped/unreachable FreeIPA

### PR DESCRIPTION
**Pull request by pdarvasi**
This commit
 - introduces FreeIPA validation during recovery() and validateRecoverable() calls on CB side
 - refactors FreeIPA validation on SDX side

### CODE REVIEW

File: core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/stacks/response/recovery/RecoveryValidationV4Response.java [hunk 1]
```diff
@@ -30,6 +35,25 @@ public void setStatus(RecoveryStatus status) {
         this.status = status;
     }
 
+    public RecoveryValidationV4Response merge(RecoveryValidationV4Response other) {
```
Comment (foldik) [inline core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/stacks/response/recovery/RecoveryValidationV4Response.java:38]: Can you please add unit tests for this?

File: core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/stacks/response/recovery/RecoveryValidationV4Response.java [hunk 2]
```diff
@@ -30,6 +35,25 @@ public void setStatus(RecoveryStatus status) {
...
+        boolean nonRecoverable = status.nonRecoverable();
+        boolean otherNonRecoverable = other.getStatus().nonRecoverable();
+        RecoveryStatus mergedState = nonRecoverable || otherNonRecoverable ? NON_RECOVERABLE : RECOVERABLE;
+        String delimiter = mergedState.recoverable() ? "" : " Next issue: ";
```
Comment (foldik) [inline core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/stacks/response/recovery/RecoveryValidationV4Response.java:42]:
```java
        System.out.println(new RecoveryValidationV4Response("Success.", RecoveryStatus.RECOVERABLE)
                .merge(new RecoveryValidationV4Response("Success.", RecoveryStatus.RECOVERABLE)).getReason());
```

This results in the message `Success.Success.`. Can you please add a white space as the delimiter and maybe later trim the result?

File: core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/stacks/response/recovery/RecoveryValidationV4Response.java [hunk 3]
```diff
@@ -30,6 +35,25 @@ public void setStatus(RecoveryStatus status) {
...
+        StringJoiner mergedReason = new StringJoiner(delimiter);
+        if (nonRecoverable) {
+            mergedReason.add(getReason());
+            if (otherNonRecoverable) {
+                mergedReason.add(other.getReason());
+            }
+        } else if (otherNonRecoverable) {
+            mergedReason.add(other.getReason());
+        } else {
+            mergedReason.add(getReason()).add(other.getReason());
```
Comment (foldik) [inline core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/stacks/response/recovery/RecoveryValidationV4Response.java:52]: Can you inline the getters?

File: core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/ClusterRecoveryService.java [hunk 4]
```diff
@@ -72,16 +85,30 @@ public RecoveryValidationV4Response validateRecovery(Long workspaceId, NameOrCrn
...
+        boolean freeIpaAvailable = freeipaService.freeipaStatusInDesiredState(stack, Set.of(Status.AVAILABLE));
+        String reason = "";
+        RecoveryStatus status;
+        if (!freeIpaAvailable) {
+            reason = "Recovery cannot be performed because the FreeIPA isn't available. Please check the FreeIPA state and try again.";
+            status = NON_RECOVERABLE;
+        } else {
+            status = RECOVERABLE;
         }
         LOGGER.info(reason);
```
Comment (foldik) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/ClusterRecoveryService.java:113]: Can we not log an empty message when freeipa is available?

File: datalake/src/main/java/com/sequenceiq/datalake/service/FreeipaService.java [hunk 5]
```diff
@@ -18,7 +19,17 @@
     @Inject
     private FreeIpaV1Endpoint freeIpaV1Endpoint;
 
-    public DescribeFreeIpaResponse describe(String envCrn) {
+    public void checkFreeipaRunning(String envCrn) {
+        DescribeFreeIpaResponse freeipa = describe(envCrn);
+        if (freeipa != null && freeipa.getAvailabilityStatus() != null && !freeipa.getAvailabilityStatus().isAvailable()) {
```
Comment (foldik) [inline datalake/src/main/java/com/sequenceiq/datalake/service/FreeipaService.java:24]: I don't get it. If the freeipa is null or the availability status is null then we don't throw exception? Can we throw an exception in those cases as well and catch the problem early on? Now it doesn't really fail fast.

File: datalake/src/test/java/com/sequenceiq/datalake/service/FreeipaServiceTest.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+import org.mockito.junit.jupiter.MockitoExtension;
+
+import com.sequenceiq.cloudbreak.common.exception.BadRequestException;
+import com.sequenceiq.freeipa.api.v1.freeipa.stack.FreeIpaV1Endpoint;
+import com.sequenceiq.freeipa.api.v1.freeipa.stack.model.common.AvailabilityStatus;
+import com.sequenceiq.freeipa.api.v1.freeipa.stack.model.common.Status;
+import com.sequenceiq.freeipa.api.v1.freeipa.stack.model.describe.DescribeFreeIpaResponse;
+
+@ExtendWith(MockitoExtension.class)
+class FreeipaServiceTest {
```
Comment (foldik) [inline datalake/src/test/java/com/sequenceiq/datalake/service/FreeipaServiceTest.java:20]: Please add tests to make sure if the required freeipa or the status is null then the validation fails as well.

**Review by foldik — CHANGES_REQUESTED** [review]
One change for validation and request for unit tests.

File: datalake/src/main/java/com/sequenceiq/datalake/service/FreeipaService.java (hunk 5 above)
Comment (pdarvasi) [reply datalake/src/main/java/com/sequenceiq/datalake/service/FreeipaService.java:24]: This is exactly the former logic present in 2 services in method `private void checkFreeipaRunning(String envCrn) {` extracted to its service. Otherwise, your suggestion is valid, so I am addressing it.

**Review by foldik — APPROVED** [review]
LGTM
